hedera-node/src/main/java/com/hedera/services/ledger/interceptors/StakeAwareAccountsCommitsInterceptor.java

```diff
@@ -0,0 +1,319 @@
+		for (int i = 0, n = pendingChanges.size(); i < n; i++) {
+			final var account = pendingChanges.entity(i);
+			final var changes = pendingChanges.changes(i);
+			final var accountNum = pendingChanges.id(i).getAccountNum();
+
+			checksFundingBalanceChange(changes, account, accountNum);
+
+			// Update BALANCE and STAKE_PERIOD_START in the pending changes for this account, if reward-eligible
+			if (isRewardable(account, changes, latestEligibleStart)) {
+				payReward(account, changes, accountNum, hasBeenRewarded);
+			}
+			// Update any STAKED_TO_ME side effects of this change
+			n = updateStakedToMeSideEffects(i, pendingChanges, hasBeenRewarded, latestEligibleStart);
+		}
+
+		// Now iterate through the change set again to update node stakes; we do this is in a
+		// separate loop to ensure all STAKED_TO_ME fields have their final values
+		updateNodeStakes(pendingChanges);
+		trackRewardsPaidByFunding(pendingChanges);
+
+		activateRewardsIfValid();
+	}
+
+	private void checksFundingBalanceChange(final Map<AccountProperty, Object> changes,
+			final MerkleAccount account, final long accountNum) {
+		if (changes.containsKey(AccountProperty.BALANCE)) {
+			final long newBalance = (long) changes.get(AccountProperty.BALANCE);
+			if (account != null && (accountNum == STAKING_FUNDING_ACCOUNT_NUMBER)) {
+				rewardBalanceChanged = true;
+				newRewardBalance = newBalance;
+			}
+		}
+	}
+
+	private void trackRewardsPaidByFunding(
+			final EntityChangeSet<AccountID, MerkleAccount, AccountProperty> pendingChanges) {
+		final var rewardsPaid = rewardCalculator.rewardsPaidInThisTxn();
+		final var rewardAccountI = findOrAdd(800L, pendingChanges);
+		manager.updateBalance(-rewardsPaid, rewardAccountI, pendingChanges);
+	}
+
+	private void updateNodeStakes(final EntityChangeSet<AccountID, MerkleAccount, AccountProperty> pendingChanges) {
+		for (int i = 0, n = pendingChanges.size(); i < n; i++) {
+			final var account = pendingChanges.entity(i);
+			final var changes = pendingChanges.changes(i);
+
+			final var curNodeId = (account != null) ? account.getStakedId() : 0L;
+			final var newNodeId = manager.getNodeStakeeNum(changes);
+			if (curNodeId < 0 && curNodeId != newNodeId) {
+				// Node stakee has been replaced, withdraw stakeRewarded or stakeNotRewarded from ex-stakee based on
+				// isDeclineReward option
+				manager.withdrawStake(
+						Math.abs(curNodeId),
+						account.getBalance() + account.getStakedToMe(),
+						manager.finalDeclineRewardGiven(account, changes));
+			}
+			if (newNodeId < 0) {
+				// Award updated stake to new node stakee to the fields stakeRewarded or stakeNotRewarded from
+				// ex-stakee based on isDeclineReward option
+				manager.awardStake(
+						Math.abs(newNodeId),
+						finalBalanceGiven(account, changes) + manager.finalStakedToMeGiven(account, changes),
+						manager.finalDeclineRewardGiven(account, changes));
+			}
+		}
+	}
+
+	int updateStakedToMeSideEffects(
+			final int num,
```

Comment:
Why not just pass in `account` and `changes`?

Reply:
Done , passed both account, changes in addition to pendingChanges and latestRewardableStart